docs/development/typescript.md

```diff
@@ -0,0 +1,56 @@
+More details are available in the [official docs](https://www.typescriptlang.org/docs/handbook/project-references.html)
+
+### Caveats
+This architecture imposes several limitations to which we must comply:
+- Projects cannot have circular dependencies. Even though the Kibana platform doesn't support circular dependencies between Kibana plugins, TypeScript (and ES6 modules) does allow circular imports between files. So in theory, you may face a problem when migrating to the TS project references and you will have to resolve this circular dependency.
+- A project must emit its type declaration. It's not always possible to generate a type declaration if the compiler cannot infer a type. There are two basic cases:
+1. Your plugin exports a type inferring an internal type declared in Kibana codebase. In this case, you'll have to either export an internal type or to declare an exported type explicitly.
+2. Your plugin exports something inferring a type from a 3rd party library that doesn't export this type. To fix the problem, you have to declare the exported type manually.
+
+### Prerequisitions
+Since `tsc` doesn't support circular project references, the migration order does matter. You can migrate your plugin only when all the plugin dependencies already have migrated. It creates a situation where commonly used plugins (such as `data` or `kibana_react`) have to migrate first.
+
+### Implementation
+- make sure all the plugins listed as dependencies in `kibana.json` file have migrated to TS project references.
+- add `tsconfig.json` in the root folder of your plugin.
+```json
+{
+  "extends": "../../../tsconfig.base.json",
+  "compilerOptions": {
+    "composite": true,
+    "outDir": "./target/types",
+    "emitDeclarationOnly": true,
+    "declaration": true,
+    "declarationMap": true
+  },
+  "include": [
+    // add all the folders containg files to be compiled
+  ],
+  "references": [
+    { "path": "../../core/tsconfig.json" },
+    // add references to other TypeScript projects your plugin dependes on
+  ]
+}
+```
+If your plugin imports a file not listed in `include`, the build will fail with the next message `File ‘…’ is not listed within the file list of project …’. Projects must list all files or use an 'include' pattern.`
+- build you plugin `./node_modules/.bin/tsc -b src/plugins/my_plugin`. Fix errors if `tsc` cannot generate type declarations for your project.
+- make sure the `target/types` folder doesn’t contain files not belonging to your project. Otherwise, it means your plugin includes files from another project by mistake.
+- add your project reference to `include` property of `tsconfig.refs.json`
+- add your plugin to `include` property and plugin folder to `exclude` property of the `tsconfig.json` it used to belong to (for example, for `src/plugins/**` it's `tsconfig.json`; for `x-pack/plugins/**` it’s `x-pack/tsconfig.json`).
```

Comment:
Is this also the `references` property instead of `include` property?

Reply:
right